docs/spec/enums.rst

```diff
@@ -0,0 +1,308 @@
+Enumerations
+============
+
+Introduction
+------------
+
+The ``Enum`` class behaves differently from other Python classes in several 
+ways that require special-case handling in type checkers. This section discusses
+the Enum behaviors that should be supported by type checkers and others which
+may be supported optionally. It is recommended that library and type stub
+authors avoid using optional behaviors because these may not be supported
+by some type checkers.
+
+
+Enum Definition
+---------------
+
+Enum classes can be defined using a "class syntax" or a "function syntax".
+The function syntax offers several ways to specify enum members: names passed
+as individual arguments, a list of names, a tuple of names, or a string of
+comma-delimited names, or a string of space-delimited names.
+
+Type checkers should support the class syntax, but the function syntax (in
+is various forms) is optional::
+
+    class Color1(Enum): # Supported
+        RED = 1
+        GREEN = 2
+        BLUE = 3
+
+    Color2 = Enum('Color2', 'RED', 'GREEN', 'BLUE') # Optional
+    Color3 = Enum('Color3', ['RED', 'GREEN', 'BLUE']) # Optional
+    Color4 = Enum('Color4', ('RED', 'GREEN', 'BLUE')) # Optional
+    Color5 = Enum('Color5', 'RED, GREEN, BLUE') # Optional
+    Color6 = Enum('Color6', 'RED GREEN BLUE') # Optional
+
+If a type checker supports the functional syntax, it should enforce name
+consistency. That is, if the type is assigned to a variable, the name of
+the variable must match the name of the enum class::
+
+  WrongName = Enum('Color', 'RED GREEN BLUE') # Type checker error
+
+
+Enum Behaviors
+--------------
+
+Enum classes are iterable and indexable, and they can be called with a value
+to look up the enum member with that value. Type checkers should support these
+behaviors::
+
+    class Color(Enum):
+        RED = 1
+        GREEN = 2
+        BLUE = 3
+
+    for color in Color:
+        reveal_type(color) # Revealed type is 'Color'
+
+    reveal_type(Color["RED"]) # Revealed type is 'Literal[Color.RED]' (or 'Color')
+    reveal_type(Color(3)) # Revealed type is 'Literal[Color.BLUE]' (or 'Color')
+
+Calling an enum class does not invoke its constructor, as is normal
+in most Python classes. The defined ``__new__`` method of an enum class is
+replaced with a custom implementation that performs a value-based lookup of
+an enum member.
+
+Enum members are immutable. Type checkers should generate an error if an enum
```

Comment:
Strictly speaking, the enumeration itself is immutable, but the members within the enum are mutable: while you cannot add or delete members from the enumeration, the members themselves support the setting of arbitrary attributes in the same way as the vast majority of "normal" Python classes

Reply:
```pycon
>>> from enum import Enum
>>> class Foo(Enum):
...   BAR = 'bar'
...   
>>> Foo.BAR.whatever = 42
>>> Foo.BAR.whatever
42
```